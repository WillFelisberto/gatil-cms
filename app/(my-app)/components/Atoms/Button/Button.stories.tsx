import type { Meta, StoryObj } from '@storybook/react';
import { FaCat } from 'react-icons/fa';

import { Button } from './Button';

const meta: Meta<typeof Button> = {
  title: 'Atoms/Button',
  component: Button,
  tags: ['autodocs'],
  args: {
    type: 'button'
  }
};

export default meta;
type Story = StoryObj<typeof Button>;

export const Dark: Story = {
  args: {
    children: 'Quem Somos',
    variant: 'dark',
    icon: <FaCat className="text-white" />,
    'data-testid': 'btn-dark'
  }
};

export const Light: Story = {
  args: {
    children: 'Apadrinhe',
    variant: 'light',
    icon: <FaCat className="text-white" />,
    'data-testid': 'btn-light'
  }
};

export const NoIcon: Story = {
  args: {
    children: 'Sem √çcone',
    variant: 'dark',
    'data-testid': 'btn-noicon'
  }
};
