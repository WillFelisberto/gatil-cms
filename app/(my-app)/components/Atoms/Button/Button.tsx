import clsx from 'clsx';
import { ButtonHTMLAttributes, ReactNode } from 'react';

type Variant = 'light' | 'dark';

type ButtonProps = {
  children: string;
  variant?: Variant;
  icon?: ReactNode;
  className?: string;
  'data-testid'?: string;
} & ButtonHTMLAttributes<HTMLButtonElement>;

export const Button = ({
  children,
  variant = 'dark',
  icon,
  className,
  'data-testid': testId,
  ...props
}: ButtonProps) => {
  return (
    <button
      type={props.type || 'button'}
      className={clsx(
        'flex items-center justify-center gap-2 cursor-pointer px-6 py-2 rounded-full font-bold transition-colors duration-200 text-sm md:text-base',
        variant === 'dark' && 'bg-blue-950 text-white hover:bg-blue-950/90',
        variant === 'light' && 'bg-blue-300 text-white hover:bg-blue-400',
        className
      )}
      data-testid={testId}
      {...props}
    >
      <span>{children}</span>
      {icon && <span className="w-5 h-5">{icon}</span>}
    </button>
  );
};
