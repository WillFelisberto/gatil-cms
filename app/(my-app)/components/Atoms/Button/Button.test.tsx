import { FaCat } from 'react-icons/fa';

import { fireEvent, render, screen } from '@/tests/test-utils';

import { Button } from './Button';

describe('<Button />', () => {
  it('Should renders with default props', () => {
    const { container } = render(<Button>Click me</Button>);
    expect(screen.getByText('Click me')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('Should renders with light variant', () => {
    render(<Button variant="light">Light</Button>);
    const button = screen.getByRole('button', { name: /light/i });
    expect(button).toHaveClass('bg-blue-300');
    expect(button).toHaveClass('text-white');
  });

  it('Should renders with dark variant (default)', () => {
    render(<Button>Dark</Button>);
    const button = screen.getByRole('button', { name: /dark/i });
    expect(button).toHaveClass('bg-blue-950');
    expect(button).toHaveClass('text-white');
  });

  it('Should applies custom className', () => {
    render(<Button className="custom-class">With Class</Button>);
    const button = screen.getByRole('button', { name: /with class/i });
    expect(button).toHaveClass('custom-class');
  });

  it('Should renders an icon when provided', () => {
    render(<Button icon={<FaCat data-testid="cat-icon" />}>Icon</Button>);
    expect(screen.getByTestId('cat-icon')).toBeInTheDocument();
  });

  it('Should uses the correct button type when provided', () => {
    render(<Button type="submit">Submit</Button>);
    const button = screen.getByRole('button', { name: /submit/i });
    expect(button).toHaveAttribute('type', 'submit');
  });

  it('Should defaults to type="button" when no type is provided', () => {
    render(<Button>Default Type</Button>);
    const button = screen.getByRole('button', { name: /default type/i });
    expect(button).toHaveAttribute('type', 'button');
  });

  it('Should supports data-testid', () => {
    render(<Button data-testid="custom-btn">Testing</Button>);
    expect(screen.getByTestId('custom-btn')).toBeInTheDocument();
  });

  it('Should calls onClick when clicked', () => {
    const handleClick = jest.fn();
    render(<Button onClick={handleClick}>Click</Button>);
    fireEvent.click(screen.getByRole('button', { name: /click/i }));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('Should renders children correctly', () => {
    render(<Button>Test Children</Button>);
    expect(screen.getByText('Test Children')).toBeInTheDocument();
  });
});
