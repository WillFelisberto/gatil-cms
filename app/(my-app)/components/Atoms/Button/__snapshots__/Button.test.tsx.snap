// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Should renders with default props 1`] = `
<div>
  <button
    class="flex items-center justify-center gap-2 cursor-pointer px-6 py-2 rounded-full font-bold transition-colors duration-200 text-sm md:text-base bg-blue-950 text-white hover:bg-blue-950/90"
    type="button"
  >
    <span>
      Click me
    </span>
  </button>
</div>
`;
